apps/btc_family/btc_txn.c

```diff
@@ -247,31 +285,183 @@ static bool fetch_transaction_meta(btc_query_t *query) {
+    // P2PK 68, P2PKH 25 (21 excluding OP_CODES), P2WPKH 22, P2MS ~, P2SH 23 (21
+    // excluding OP_CODES) refer https://learnmeabitcoin.com/technical/script
+    // for explaination Currently the device can spend P2PKH or P2WPKH inputs
+    // only for (int i = 0; i < in_count; i++) {
+    //   if (txn_ctx->inputs[i].script_pub_key.size != 22 &&
+    //       txn_ctx->inputs[i].script_pub_key.size != 25) {
+    //     return false;
+    //   }
+    // }
+
+    // verify transaction details and discard the raw-transaction (prev_txn)
+    const btc_sign_txn_input_prev_txn_t *txn = &query->sign_txn.input.prev_txn;
+    if (!btc_verify_input(
+            txn->bytes, txn->size, &btc_txn_context->inputs[idx])) {
+      // input validation failed, terminate immediately
+      btc_send_error(ERROR_COMMON_ERROR_CORRUPT_DATA_TAG,
+                     ERROR_DATA_FLOW_INVALID_DATA);
+      return false;
+    }
+
+    // clone the input details into btc_txn_context
+    btc_txn_input_t *input = &btc_txn_context->inputs[idx];
+    input->prev_output_index = query->sign_txn.input.prev_output_index;
+    input->address_index = query->sign_txn.input.address_index;
+    input->change_index = query->sign_txn.input.change_index;
+    input->value = query->sign_txn.input.value;
+    input->sequence = query->sign_txn.input.sequence;
+    input->script_pub_key.size = query->sign_txn.input.script_pub_key.size;
+    memcpy(input->prev_txn_hash, query->sign_txn.input.prev_txn_hash, 32);
+    memcpy(input->script_pub_key.bytes,
+           query->sign_txn.input.script_pub_key.bytes,
+           input->script_pub_key.size);
+
+    // send accepted response to indicate validation of input passed
+    send_response(BTC_SIGN_TXN_RESPONSE_INPUT_ACCEPTED_TAG);
+  }
+  return true;
+}
+
+static bool fetch_output(btc_query_t *query) {
+  // track if it is a zero valued transaction; all input is going into fee
+  bool zero_value_transaction = true;
+
+  for (int idx = 0; idx < btc_txn_context->metadata.output_count; idx++) {
+    if (!btc_get_query(query, BTC_QUERY_SIGN_TXN_TAG) &&
+        !check_which_request(query, BTC_SIGN_TXN_REQUEST_OUTPUT_TAG)) {
+      return false;
+    }
+
+    btc_sign_txn_output_t *output = &query->sign_txn.output;
+    memcpy(
+        &btc_txn_context->outputs[idx], output, sizeof(btc_sign_txn_output_t));
+    if (0 != output->value) {
+      zero_value_transaction = false;
+    }
+    if (OP_RETURN == output->script_pub_key.bytes[0] && 0 != output->value) {
+      // ensure any funds are not being locked (not spendable)
+      btc_send_error(ERROR_COMMON_ERROR_CORRUPT_DATA_TAG,
+                     ERROR_DATA_FLOW_INVALID_DATA);
+      return false;
+    }
+  }
+  if (true == zero_value_transaction) {
+    // do not allow zero valued transaction; all input is going into fee
+    btc_send_error(ERROR_COMMON_ERROR_CORRUPT_DATA_TAG,
+                   ERROR_DATA_FLOW_INVALID_DATA);
+    return false;
+  }
+  return true;
+}
```

Comment:
I think we need to send some data to SDK here.

Reply:
added